admission: error accounting for disk reads and writes

aadityasondhi opened the pull request:
Previously, we would ignore errors in disk token accounting hoping that
the next adjustment interval would capture them. This caused issues when
async compactions ran within an adjustment interval. These could spike
the bandwidth for a section of the 15 second interval without moving the
average too much (write-amp estimation).

This patch introduces a higher frequency error accounting mechanism. It
runs every 1s to account for any additional reads and writes that could
have occurred. Errors are only deducted if the read and/or write
bandwidth in the interval is greater than the number of tokens already
deducted in that interval.

Fixes #132332.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133310)
<!-- Reviewable:end -->

aadityasondhi commented:
Changes in `pkg/util/admission/testdata/io_load_listener` are merely a rewrite due to the additional `elasticDiskReadTokens` and `elasticDiskReadTokensAllocated` fields.

sumeerbhola reviewed (changes requested):
sorry for the delay

> Reviewed 4 of 10 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-OBXGLMQ03DLOthH-swy:bdpui8l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @itsbilal)

___
*[`pkg/util/admission/granter.go` line 306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXBVe64peVJCEy4zW9:-OBXBVe64peVJCEy4zWA:b-q9ih6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L306)):*
> ```Go
> 		diskTokensAvailable diskTokens
> 		diskTokensError     struct {
> 			prevObservedWrites             uint64
> ```

Can you add a comment
// prevObserved{Writes,Reads} is the observed disk metrics in the last call to adjustDiskTokenErrorLocked. These are used to compute the delta.
___
*[`pkg/util/admission/granter.go` line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXBiB0B_-XDPmAd18p:-OBXBiB0B_-XDPmAd18q:bnfww38) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L309)):*
> ```Go
> 			prevObservedReads              uint64
> 			diskWriteTokensAlreadyDeducted int64
> 			diskReadTokensAlreadyDeducted  int64
> ```

I realize this will change in the future, but I think it is important to document what these are now.

// diskWriteTokensAlreadyDeducted are the deductions we have seen over the last error adjustment interval.

// diskReadTokensAlreadyDeducted are the expected deductions over an error adjustment interval. We compute the expected deductions once every 15s and then dole them out in setAvailableTokens.

I think there is a problem here. Say we observed R reads over the last 15s interval. The `elasticDiskReadTokensCapacity` will be R/60. But our error adjustment interval is 1s. So even if there is no error, the observed reads will be R/15 and we will deduct more. The easiest fix would be to make the error adjustment interval equal to 250ms so that these align. And write a detailed code comment on why we are aligning these.

Also, there may be a bit of misalignment risk with the `shouldAdjustForError()` logic working independently of the ticks. If the error interval is 250ms and it fires at 3ms and 253ms, while the `diskReadTokensAlreadyDeducted` was also being updated at the `unloadedDuration` of 250ms and got written at 2ms and 254ms, then the second error adjustment will find `diskReadTokensAlreadyDeducted` equal to zero. I don't think the write tokens suffer from this problem since the deductions are live deductions. We may need to rework the `tokenAllocatorTicker` logic to introduce coordination. 
___
*[`pkg/util/admission/granter.go` line 522 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXFcpp0x5eW1xCFkMu:-OBXFcpp0x5eW1xCFkMv:bd5mvke) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L522)):*
> ```Go
> 	if writeError > 0 {
> 		if sg.coordMu.diskTokensAvailable.writeByteTokens < 0 {
> 			sg.coordMu.diskTokensAvailable.writeByteTokens = 0
> ```

I suspect this is motivated by https://docs.google.com/document/d/1KelFCIUd9jaBkAev5G6CReI\_kA-u\_pv6CHHNjjD4-lk/edit?tab=t.0#bookmark=id.czsnhcjsmyia

I suspect the approach mentioned in that doc is not quite right. It is possible that `writeByteTokens` are negative because of a huge write and not due to previous error correction. This needs some more thought.
___
*[`pkg/util/admission/granter.go` line 538 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXGDKwA0OhSZMc1H-X:-OBXGDKwA0OhSZMc1H-Y:b-puri8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L538)):*
> ```Go
> 		if sg.coordMu.diskTokensAvailable.writeByteTokens < 0 {
> 			sg.coordMu.diskTokensAvailable.writeByteTokens = 0
> 		}
> ```

why this resetting to 0 after subtracting the `readError`?
___
*[`pkg/util/admission/io_load_listener.go` line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBW8eauBXLeUKZI_M_K:-OBW8eauBXLeUKZI_M_L:b-6ttgta) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/io_load_listener.go#L249)):*
> ```Go
> 	totalNumElasticByteTokens  int64
> 	elasticByteTokensAllocated int64
> ```

I think we should add a comment stating that even though these are prefixed with elastic, these are tokens for both regular and elastic work. These are only prefixed with elastic as only elastic work waits for them. Technically, elastic reads don't wait for elasticDiskReadTokens either, so perhaps we should just drop the elastic prefix on these fields.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Thanks for the review. I haven't adressed anything, just want to discuss the synchronization thing before I make changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-OBXQJ-N2H0GLKTK5DsW:b-aci8wc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/util/admission/granter.go` line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXBiB0B_-XDPmAd18p:-OBXLwUa8DfL5T8D0QPJ:bhtf7br) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L309)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I realize this will change in the future, but I think it is important to document what these are now.

// diskWriteTokensAlreadyDeducted are the deductions we have seen over the last error adjustment interval.

// diskReadTokensAlreadyDeducted are the expected deductions over an error adjustment interval. We compute the expected deductions once every 15s and then dole them out in setAvailableTokens.

I think there is a problem here. Say we observed R reads over the last 15s interval. The `elasticDiskReadTokensCapacity` will be R/60. But our error adjustment interval is 1s. So even if there is no error, the observed reads will be R/15 and we will deduct more. The easiest fix would be to make the error adjustment interval equal to 250ms so that these align. And write a detailed code comment on why we are aligning these.

Also, there may be a bit of misalignment risk with the `shouldAdjustForError()` logic working independently of the ticks. If the error interval is 250ms and it fires at 3ms and 253ms, while the `diskReadTokensAlreadyDeducted` was also being updated at the `unloadedDuration` of 250ms and got written at 2ms and 254ms, then the second error adjustment will find `diskReadTokensAlreadyDeducted` equal to zero. I don't think the write tokens suffer from this problem since the deductions are live deductions. We may need to rework the `tokenAllocatorTicker` logic to introduce coordination.
</blockquote></details>

Re: The error coordination issue you brought up. I was initially debating the synchronization vs no synchronization thing. I landed on this because I think it works fine (but I see a bug now after rethinking). Here is my thought processes. And let's disregard what the code currently does since I think there is a minor issue there.

1. let `I` represent the allocation interval, let `J` represent the error accounting interval
2. let's assume we have decided to deduct 100*60 (6000B) for this interval.

I0 = 0
I1 = 100
I2 = 200
I3 = 300
I4 = 400
J1 => we find 300B read from disk over the interval. We subtract and set to 0. We don't deduct additional tokens since 300 < 400.
I5 = 100
I6 = 200
I7 = 300
I8 = 400
J2 => we find 600B read from disk over the interval. We subtract and set to 0. We deduct 200 additional tokens.

The issue in my code being I currently don't let `diskReadTokensAlreadyDeducted` go above the hypothetical 100B here on every allocation tick. If I remove that limitation, I think this should work. The issue is in `pkg/util/admission/granter.go:703`:

```
sg.coordMu.diskTokensError.diskReadTokensAlreadyDeducted += elasticDiskReadTokens
	if sg.coordMu.diskTokensError.diskReadTokensAlreadyDeducted > elasticDiskReadTokensCapacity {
		sg.coordMu.diskTokensError.diskReadTokensAlreadyDeducted = elasticDiskReadTokensCapacity
	}
```

With the setup described above, I think regardless on when J ticks, we should still be accurately adjusting the error. It basically works the same way as writes do, but the deduction (which happens at admission for writes) happens during token allocation, and doesn't need to be limited to the `capacity` i.e. we keep allocating until the error accounting tick adjusts for the observed error. 

I could be missing something and it might be helpful to discuss with a whiteboard in-person.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @itsbilal)

___
*[`pkg/util/admission/granter.go` line 522 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXFcpp0x5eW1xCFkMu:-OBXcRe16dmFGqz_jvcf:b8ef17x) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L522)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I suspect this is motivated by https://docs.google.com/document/d/1KelFCIUd9jaBkAev5G6CReI\_kA-u\_pv6CHHNjjD4-lk/edit?tab=t.0#bookmark=id.czsnhcjsmyia

I suspect the approach mentioned in that doc is not quite right. It is possible that `writeByteTokens` are negative because of a huge write and not due to previous error correction. This needs some more thought.
</blockquote></details>

I was thinking a bit more about this, and trying to abstract out the problem. Assume for a moment that the error correction interval is 250ms. So we have 60 such errors to forecast. Let's call these e_f(0), ..., e_f(59). There is also the actual observed error for an interval at e_o(0), ..., e_0(59). We have to forecast e_f(i) and then 250ms later we observe e_o(i).
Roughly what we want is to:
minimize  `|e_o(i)-e_f(i)|` for all i.
Additionally, we want `e_f(i) >= 0` (assuming we are subtracting the error from tokens), since we want to err on the side of having slightly fewer tokens than to give out more tokens.

What we are currently doing is setting `e_f(i) = max(0, e_o(i-1))`. And for e_f(0), using the last observed error from the previous 60 ticks.

The idea from that doc was to additionally correct previous mistakes. In some sense trying to reduce `|sum_i e_o(i) - sum_i e_f(i)|`. This can be in opposition to the previously stated minimization goal. And the hack suggested in the doc isn't quite the right way to correct either. So I am leaning towards thinking that we shouldn't bother with this at all. That is, just subtract e_f(i) and don't care if the accumulation of error forecast have made tokens too few. But we should use the `lastTick` logic in `setAvailableTokens` to also adjust `diskTokensAvailable.writeByteTokens`, so that every 15s we restore it to 0.


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
Just to summarize what we discussed in-person re: the error accounting scheme.

1. We want to re-use the existing ticking used for token allocation to now also handle error accounting. This helps with synchronizing the two, and avoiding issues when we reset tokens every 15s.
2. We will remove the max capacity during allocation for reads. And let the error accounting deal with it.
3. We will be resetting read tokens every 15s on the last tick.
4. We will stop resetting write tokens on every error correction and instead only do it at the last tick.
5. Consider lowering the error accounting interval to be lower (maybe 250ms).

aadityasondhi committed:
admission: error accounting for disk reads and writes

Previously, we would ignore errors in disk token accounting hoping that
the next adjustment interval would capture them. This caused issues when
async compactions ran within an adjustment interval. These could spike
the bandwidth for a section of the 15 second interval without moving the
average too much (write-amp estimation).

This patch introduces a higher frequency error accounting mechanism. It
runs every 1s to account for any additional reads and writes that could
have occurred. Errors are only deducted if the read and/or write
bandwidth in the interval is greater than the number of tokens already
deducted in that interval.

Fixes #132332.

Release note: None

aadityasondhi reviewed (commented):
Resolved your outstanding comments. And changed the implementation to reflect what we discussed in person. This should ready for a final review.

PTAL @sumeerbhola 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-OC-expS3Z1iJsxjz0Kp:b-j1mxsh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/util/admission/granter.go` line 306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXBVe64peVJCEy4zW9:-OC-eeRm9ddiAIBPzi-k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L306)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add a comment
// prevObserved{Writes,Reads} is the observed disk metrics in the last call to adjustDiskTokenErrorLocked. These are used to compute the delta.
</blockquote></details>

Done.
___
*[`pkg/util/admission/granter.go` line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXBiB0B_-XDPmAd18p:-OC-eftqAt7ZWOnJoFCh:b-sq6umo) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L309)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Re: The error coordination issue you brought up. I was initially debating the synchronization vs no synchronization thing. I landed on this because I think it works fine (but I see a bug now after rethinking). Here is my thought processes. And let's disregard what the code currently does since I think there is a minor issue there.

1. let `I` represent the allocation interval, let `J` represent the error accounting interval
2. let's assume we have decided to deduct 100*60 (6000B) for this interval.

I0 = 0
I1 = 100
I2 = 200
I3 = 300
I4 = 400
J1 => we find 300B read from disk over the interval. We subtract and set to 0. We don't deduct additional tokens since 300 < 400.
I5 = 100
I6 = 200
I7 = 300
I8 = 400
J2 => we find 600B read from disk over the interval. We subtract and set to 0. We deduct 200 additional tokens.

The issue in my code being I currently don't let `diskReadTokensAlreadyDeducted` go above the hypothetical 100B here on every allocation tick. If I remove that limitation, I think this should work. The issue is in `pkg/util/admission/granter.go:703`:

```
sg.coordMu.diskTokensError.diskReadTokensAlreadyDeducted += elasticDiskReadTokens
	if sg.coordMu.diskTokensError.diskReadTokensAlreadyDeducted > elasticDiskReadTokensCapacity {
		sg.coordMu.diskTokensError.diskReadTokensAlreadyDeducted = elasticDiskReadTokensCapacity
	}
```

With the setup described above, I think regardless on when J ticks, we should still be accurately adjusting the error. It basically works the same way as writes do, but the deduction (which happens at admission for writes) happens during token allocation, and doesn't need to be limited to the `capacity` i.e. we keep allocating until the error accounting tick adjusts for the observed error. 

I could be missing something and it might be helpful to discuss with a whiteboard in-person.
</blockquote></details>

Changed implementation to reflect what we discussed.
___
*[`pkg/util/admission/granter.go` line 522 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXFcpp0x5eW1xCFkMu:-OC-en99645S11gwVL9B:b-sq6umo) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L522)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I was thinking a bit more about this, and trying to abstract out the problem. Assume for a moment that the error correction interval is 250ms. So we have 60 such errors to forecast. Let's call these e_f(0), ..., e_f(59). There is also the actual observed error for an interval at e_o(0), ..., e_0(59). We have to forecast e_f(i) and then 250ms later we observe e_o(i).
Roughly what we want is to:
minimize  `|e_o(i)-e_f(i)|` for all i.
Additionally, we want `e_f(i) >= 0` (assuming we are subtracting the error from tokens), since we want to err on the side of having slightly fewer tokens than to give out more tokens.

What we are currently doing is setting `e_f(i) = max(0, e_o(i-1))`. And for e_f(0), using the last observed error from the previous 60 ticks.

The idea from that doc was to additionally correct previous mistakes. In some sense trying to reduce `|sum_i e_o(i) - sum_i e_f(i)|`. This can be in opposition to the previously stated minimization goal. And the hack suggested in the doc isn't quite the right way to correct either. So I am leaning towards thinking that we shouldn't bother with this at all. That is, just subtract e_f(i) and don't care if the accumulation of error forecast have made tokens too few. But we should use the `lastTick` logic in `setAvailableTokens` to also adjust `diskTokensAvailable.writeByteTokens`, so that every 15s we restore it to 0.
</blockquote></details>

Changed implementation to reflect what we discussed.
___
*[`pkg/util/admission/granter.go` line 538 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBXGDKwA0OhSZMc1H-X:-OC-epCQBO2B4oNP2DlM:b-sq6umo) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/granter.go#L538)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why this resetting to 0 after subtracting the `readError`?
</blockquote></details>

Changed implementation to reflect what we discussed.
___
*[`pkg/util/admission/io_load_listener.go` line 249 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OBW8eauBXLeUKZI_M_K:-OC-esXd-kmu8A4kaqGg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ac944d4ef0ff1e36c95ace474d9e680a5ab49f/pkg/util/admission/io_load_listener.go#L249)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think we should add a comment stating that even though these are prefixed with elastic, these are tokens for both regular and elastic work. These are only prefixed with elastic as only elastic work waits for them. Technically, elastic reads don't wait for elasticDiskReadTokens either, so perhaps we should just drop the elastic prefix on these fields.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @itsbilal)

___
*[`pkg/util/admission/io_load_listener.go` line 462 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC6CEZkBEdaGo-vfrrp:-OC6CEZkBEdaGo-vfrrq:b-bk0i4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/io_load_listener.go#L462)):*
> ```Go
> 		tickDur = loadedDuration
> 	}
> 	errorTickThreshold := uint64((int64(errorAdjustmentDuration) * tickDur.ticksInAdjustmentInterval()) / int64(adjustmentInterval*time.Second))
> ```

I don't think this is robust to delayed ticks in the way that `remainingTicks` is robust.

If we are ticking at 1ms, the `errorTickThreshold=1000`. Which is reasonable since we expect 15000 ticks under perfect conditions. But we can easily jump from 14002 remaining ticks to 13998 remaining ticks (making something up). The caller of `remainingTicks` adjusts to these jumps by apportioning the remaining tokens over the remaining ticks. But the logic here is expecting the modulus to be 0. But since we never say remainingTicks=14000 it will not be 0, even though we crossed 14000. I think we need some additional state in `tokenAllocationTicker` to compensate for these jumps.
___
*[`pkg/util/admission/granter.go` line 699 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC68XOI3zabNT35fzya:-OC68XOI3zabNT35fzyb:b-yci94b) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/granter.go#L699)):*
> ```Go
> 	// NB: We don't cap the disk read tokens as they are only deducted during the
> 	// error accounting loop. So essentially, we give reads the "burst" capacity
> 	// of the error accounting interval.
> ```

Say these ticks are happening every 250ms (to make the example more concise), and each tick is adding 1MiB (i.e., diskReadTokens=1MiB). So at 1s we are at 4MiB. Then say the observed reads is 1MiB. So the `diskReadTokensAlreadyDeducted=3MiB`, and this repeats for the first 10s, which gives us `diskReadTokensAlreadyDeducted=30MiB`. So we've accumulated a "burst capacity" of 30MiB. If the next 1s sees 10MiB of reads, the burst capacity can absorb multiple seconds of such behavior, without deducting it from the write tokens.
This seems wrong. The provisioned disk bandwidth, especially on cloud virtual disks, doesn't have such multi-second burst capacity. Enforcement of the provisioned bandwidth/IOPS can happen at a fine granularity (some notes from a meeting suggest 1s interval in EBS).

We may need to go back to the idea of making the error accounting run at 250ms and use a read bandwidth capacity like the others.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Haven't made any changes. Will wait on agreement on the error ticking logic.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-OCAe_nIDEYUGLp_mko-:b-x69ppb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/util/admission/granter.go` line 699 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC68XOI3zabNT35fzya:-OCAbnJeBCFpFzXRS4Pa:b-ypjf7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/granter.go#L699)):*
That large burst capacity thing would not happen. That is because we reset `diskReadTokensAlreadyDeducted` every time we adjust error. See `adjustDiskTokenErrorLocked`. Maybe worth adding a comment here. So at most, we have a burst capacity of 1s (error interval). It has worked fine in experiments, granted I was not specifically looking for this type of issue.

> We may need to go back to the idea of making the error accounting run at 250ms and use a read bandwidth capacity like the others.

We could do this but it would also need to adapt to the 1ms ticking, otherwise we can't add a cap to the read tokens at 250ms. This would mean basically account for error, every allocation  loop. I am worried that 1ms might be too short of an interval to account for reads and writes and we may start doing too much.

We could alternatively, decouple the allocation of read tokens from the rest and allocate them at a fixed 250ms interval, which would match the error accounting interval, but it is pretty much the same thing as what we do now (1s burst). But it make this code more confusing.

I can work on making this code easier to digest, add more documentation etc., if you think the logic I described in my first statement is sound.
___
*[`pkg/util/admission/io_load_listener.go` line 462 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC6CEZkBEdaGo-vfrrp:-OCAadFgB8lzODcxhZEv:bx9x93t) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/io_load_listener.go#L462)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't think this is robust to delayed ticks in the way that `remainingTicks` is robust.

If we are ticking at 1ms, the `errorTickThreshold=1000`. Which is reasonable since we expect 15000 ticks under perfect conditions. But we can easily jump from 14002 remaining ticks to 13998 remaining ticks (making something up). The caller of `remainingTicks` adjusts to these jumps by apportioning the remaining tokens over the remaining ticks. But the logic here is expecting the modulus to be 0. But since we never say remainingTicks=14000 it will not be 0, even though we crossed 14000. I think we need some additional state in `tokenAllocationTicker` to compensate for these jumps.
</blockquote></details>

Nice callout. I did not consider that. I will add a field for last error accounting "tick". 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> Reviewed 1 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @itsbilal)

___
*[`pkg/util/admission/granter.go` line 699 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC68XOI3zabNT35fzya:-OCEAEsv1POcRDAZnZD_:b-jt8pm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/granter.go#L699)):*
> That large burst capacity thing would not happen. That is because we reset `diskReadTokensAlreadyDeducted` every time we adjust error. See `adjustDiskTokenErrorLocked`. Maybe worth adding a comment here. So at most, we have a burst capacity of 1s (error interval).

Ah yes. Definitely worth a code comment. It's easy to forget how these things interlock.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Should be good for another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-OCForqx5KyNoI4adGLe:b-v9cq43)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/util/admission/granter.go` line 699 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC68XOI3zabNT35fzya:-OCFoj6u7dmrvTIWNZG-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/granter.go#L699)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> That large burst capacity thing would not happen. That is because we reset `diskReadTokensAlreadyDeducted` every time we adjust error. See `adjustDiskTokenErrorLocked`. Maybe worth adding a comment here. So at most, we have a burst capacity of 1s (error interval).

Ah yes. Definitely worth a code comment. It's easy to forget how these things interlock.
</blockquote></details>

Done.
___
*[`pkg/util/admission/io_load_listener.go` line 462 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OC6CEZkBEdaGo-vfrrp:-OCFoman3VvYYg-pH8DG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b299d71590409cff47e883008ce2a6aa765597/pkg/util/admission/io_load_listener.go#L462)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Nice callout. I did not consider that. I will add a field for last error accounting "tick".
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 10 files at r2, 4 of 7 files at r4, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @itsbilal)

___
*[`pkg/util/admission/grant_coordinator.go` line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdEqrq2dNgkSX4fz20:-OCdEqrq2dNgkSX4fz21:b-vl51wx) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/grant_coordinator.go#L148)):*
> ```Go
> 							storeLoaded := gc.pebbleMetricsTick(ctx, m)
> 							// TODO(aaditya): Verify we go back to the unloaded state. I don't
> 							// think we do.
> ```

Good catch. But given the pain 250ms ticks have caused for CPU, I think we should leave the accidental current behavior and simply change the comment where `systemLoaded` is declared saying that we start as unloaded mainly for tests, and do a one-way transition to do 1ms ticks once load happens.
___
*[`pkg/util/admission/testdata/granter` line 837 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdUvKq3dyuC7BycxJ6:-OCdUvKq3dyuC7BycxJ7:b-caoepb) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L837)):*
> ```
> 
> # Tokens already deducted is 0. Any writes that occur will be considered error.
> adjust-disk-error actual-write-bytes=10 actual-read-bytes=10
> ```

did this deduct 20 from disk-write-tokens-avail? Seems to have deducted 10.
___
*[`pkg/util/admission/testdata/granter` line 845 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdU29f4h3JaXAMBV3d:-OCdU29f4h3JaXAMBV3e:b-iids5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L845)):*
> ```
> ----
> GrantCoordinator:
> (chain: id: 0 active: false index: 5) io-avail: 50(50), disk-write-tokens-avail: 70
> ```

can we also print the 20 for the disk read tokens?
___
*[`pkg/util/admission/testdata/granter` line 849 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdUUXV0S_Hye7estW7:-OCdUUXV0S_Hye7estW8:bfaht34) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L849)):*
> ```
> # Tokens already deducted is 0. Any writes that occur will be considered error.
> # We accounted for 20B of reads, so reads in excess of that will be error. err =
> # (10-0) + (40-20) = 30. We expect 70-30=40 tokens available.
> ```

I didn't understand the equation. Disk read error = 20 -50 = -30
Disk write error = 0 - 20 = -20.
So we have a total error of -50? Should that make the avail tokens = 70-50?
___
*[`pkg/util/admission/io_load_listener.go` line 469 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdIGM21ly-MJ1Gc3Vb:-OCdIGM21ly-MJ1Gc3Vc:bcjnd9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/io_load_listener.go#L469)):*
> ```Go
> 		t.lastErrorAdjustmentTick = uint64(tickDur.ticksInAdjustmentInterval())
> 	}
> 	errorTickThreshold := uint64((int64(errorAdjustmentDuration) * tickDur.ticksInAdjustmentInterval()) / int64(adjustmentInterval*time.Second))
> ```

I think it would be good to separate the const part of this computation into a declared const for `(adjustmentInterval*time.Second)/errorAdjustmentDuration`. It makes it easier to understand.

Also, an example of how t.lastErrorAdjustmentTick will evolve over an adjustmentInterval would help the reader. It took me some time to understand what was going on here.
___
*[`pkg/util/admission/io_load_listener.go` line 475 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdJ8zx6PxqzFsKpnj3:-OCdJ8zx6PxqzFsKpnj4:bm7xxsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/io_load_listener.go#L475)):*
> ```Go
> 	if !shouldAdjust {
> 		return false
> 	}
> ```


Is `TestTokenAllocationTicker` updated to test the new logic?


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Sorry for taking this long in getting back to this. Still have one more thing to address.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-ODw6mNW10wZXgfur08n:bmhq8g8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/util/admission/grant_coordinator.go` line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdEqrq2dNgkSX4fz20:-ODvvhJJAWBsRRSNs3ly:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/grant_coordinator.go#L148)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Good catch. But given the pain 250ms ticks have caused for CPU, I think we should leave the accidental current behavior and simply change the comment where `systemLoaded` is declared saying that we start as unloaded mainly for tests, and do a one-way transition to do 1ms ticks once load happens.
</blockquote></details>

Done.
___
*[`pkg/util/admission/io_load_listener.go` line 469 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdIGM21ly-MJ1Gc3Vb:-ODvvo5d82Uvukuq4dLN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/io_load_listener.go#L469)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think it would be good to separate the const part of this computation into a declared const for `(adjustmentInterval*time.Second)/errorAdjustmentDuration`. It makes it easier to understand.

Also, an example of how t.lastErrorAdjustmentTick will evolve over an adjustmentInterval would help the reader. It took me some time to understand what was going on here.
</blockquote></details>

Done.
___
*[`pkg/util/admission/io_load_listener.go` line 475 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdJ8zx6PxqzFsKpnj3:-ODvvpBy3KJe0qf3t5Qw:b-w1fass) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/io_load_listener.go#L475)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is `TestTokenAllocationTicker` updated to test the new logic?
</blockquote></details>

Nice catch. Added a separate test. Look at `TestTokenAllocationTickerErrorAdjustmentThreshold`.
___
*[`pkg/util/admission/testdata/granter` line 837 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdUvKq3dyuC7BycxJ6:-ODvx7zS43FJ9Iqx0Ams:b-us90yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L837)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

did this deduct 20 from disk-write-tokens-avail? Seems to have deducted 10.
</blockquote></details>

Since this was the first iteration. We had `unlimited` read tokens. And because of that we don't deduct them. This is a little unfortunate setup, since I wanted to use the `unlimited` value to denote "unset" tokens.

Now that I think of it, maybe we should just leave it at 0 if unset since disk write tokens will be unlimited. I will put up a change in a bit to address that, flushing what I have for now.
___
*[`pkg/util/admission/testdata/granter` line 845 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdU29f4h3JaXAMBV3d:-ODw63V0COlUNCOZvB9h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L845)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can we also print the 20 for the disk read tokens?
</blockquote></details>

Done.
___
*[`pkg/util/admission/testdata/granter` line 849 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdUUXV0S_Hye7estW7:-ODvw9SI4mPlDeKDIEzO:b-5xti0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L849)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I didn't understand the equation. Disk read error = 20 -50 = -30
Disk write error = 0 - 20 = -20.
So we have a total error of -50? Should that make the avail tokens = 70-50?
</blockquote></details>

I added more commentary to explain what happened here.

Basically, the bytes passed into the function are cumulative, so we look at the delta. The delta here is, write = 10, read = 40.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Should be ready for another look. Thanks for your patience with this one!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-ODwf5L1Cxr60qEkTDM3:byrqn8c)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/util/admission/testdata/granter` line 837 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-OCdUvKq3dyuC7BycxJ6:-ODwezA-6ctBleOZBda-:b-h63bqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/61eea9fbccadbab013e7d5daab796e357e1eca87/pkg/util/admission/testdata/granter#L837)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Since this was the first iteration. We had `unlimited` read tokens. And because of that we don't deduct them. This is a little unfortunate setup, since I wanted to use the `unlimited` value to denote "unset" tokens.

Now that I think of it, maybe we should just leave it at 0 if unset since disk write tokens will be unlimited. I will put up a change in a bit to address that, flushing what I have for now.
</blockquote></details>

Done.

I set the default value for these tokens to be 0 rather than unlimited. Added some commentary for the reasoning of doing so. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 7 files at r4, 3 of 5 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133310#-:-OE10sUM9tzUSEaD5Iv3:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @itsbilal)



<!-- Sent from Reviewable.io -->

aadityasondhi commented:
TFTR! 

bors r=sumeerbhola
